File: Sources/CococoJinja/Expression/Filters.swift
//
//  Filters.swift
//  CococoJinja
//
//  Copyright Triple C Labs GmbH
//
//  The standard filter allowlist. Each filter is type-checked and throws a typed
//  ExpressionError on misuse. This is the entire callable surface of the dialect —
//  there are no globals and no user-defined filters. See
//  docs/workflow-expression-dialect.md §8.
//

import Foundation

extension FilterRegistry {
    /// Pulse's standard filter set: scalar/collection helpers plus the structured
    /// filters that absorb the old `$op` transform DSL.
    public static let standard = FilterRegistry([
        // Scalar / collection
        "default": Filters.default,
        "length": Filters.length,
        "min": Filters.min,
        "max": Filters.max,
        "lower": Filters.lower,
        "upper": Filters.upper,
        "trim": Filters.trim,
        "join": Filters.join,
        "abs": Filters.abs,
        "round": Filters.round,
        // Structured (transform DSL replacements)
        "merge": Filters.merge,
        "pick": Filters.pick,
        "omit": Filters.omit,
        "concat": Filters.concat,
    ])
}

enum Filters {

    // MARK: - Argument helpers

    private static func requireArgs(_ args: [JinjaValue], _ count: Int, _ name: String) throws {
        guard args.count == count else {
            throw ExpressionError.evaluate("filter '\(name)' expects \(count) argument(s), got \(args.count)")
        }
    }

    private static func requireString(_ value: JinjaValue, _ name: String) throws -> String {
        guard case let .string(s) = value else {
            throw ExpressionError.evaluate("filter '\(name)' requires a string, got \(value.typeName)")
        }
        return s
    }

    private static func requireObject(_ value: JinjaValue, _ name: String) throws -> [String: JinjaValue] {
        guard case let .object(o) = value else {
            throw ExpressionError.evaluate("filter '\(name)' requires an object, got \(value.typeName)")
        }
        return o
    }

    private static func requireKeyArgs(_ args: [JinjaValue], _ name: String) throws -> [String] {
        try args.map { try requireString($0, name) }
    }

    // MARK: - Scalar / collection filters

    /// `value | default(fallback)` — fallback when the input is null.
    static let `default`: ExpressionFilter = { input, args in
        try requireArgs(args, 1, "default")
        if case .null = input { return args[0] }
        return input
    }

    /// `value | length` — element/character count of an array, object, or string.
    static let length: ExpressionFilter = { input, args in
        try requireArgs(args, 0, "length")
        switch input {
        case let .array(a): return .int(a.count)
        case let .object(o): return .int(o.count)
        case let .string(s): return .int(s.count)
        default:
            throw ExpressionError.evaluate("filter 'length' requires an array, object, or string, got \(input.typeName)")
        }
    }

    /// `array | min` — smallest element by ordered comparison.
    static let min: ExpressionFilter = { input, args in
        try requireArgs(args, 0, "min")
        return try extreme(input, name: "min", keepWhen: .orderedAscending)
    }

    /// `array | max` — largest element by ordered comparison.
    static let max: ExpressionFilter = { input, args in
        try requireArgs(args, 0, "max")
        return try extreme(input, name: "max", keepWhen: .orderedDescending)
    }

    private static func extreme(_ input: JinjaValue, name: String, keepWhen wanted: ComparisonResult) throws -> JinjaValue {
        guard case let .array(items) = input else {
            throw ExpressionError.evaluate("filter '\(name)' requires an array, got \(input.typeName)")
        }
        guard var best = items.first else {
            throw ExpressionError.evaluate("filter '\(name)' requires a non-empty array")
        }
        for item in items.dropFirst() where try JinjaValue.semanticCompare(item, best) == wanted {
            best = item
        }
        return best
    }

    /// `string | lower` / `upper` / `trim`.
    static let lower: ExpressionFilter = { input, args in
        try requireArgs(args, 0, "lower")
        return .string(try requireString(input, "lower").lowercased())
    }
    static let upper: ExpressionFilter = { input, args in
        try requireArgs(args, 0, "upper")
        return .string(try requireString(input, "upper").uppercased())
    }
    static let trim: ExpressionFilter = { input, args in
        try requireArgs(args, 0, "trim")
        return .string(try requireString(input, "trim").trimmingCharacters(in: .whitespacesAndNewlines))
    }

    /// `array | join(sep)` — render elements to strings and join (default sep "").
    static let join: ExpressionFilter = { input, args in
        guard case let .array(items) = input else {
            throw ExpressionError.evaluate("filter 'join' requires an array, got \(input.typeName)")
        }
        let sep: String
        switch args.count {
        case 0: sep = ""
        case 1: sep = try requireString(args[0], "join")
        default: throw ExpressionError.evaluate("filter 'join' expects 0 or 1 argument(s), got \(args.count)")
        }
        return .string(items.map { $0.renderedString() }.joined(separator: sep))
    }

    /// `number | abs`.
    static let abs: ExpressionFilter = { input, args in
        try requireArgs(args, 0, "abs")
        switch input {
        case let .int(i):
            // `Swift.abs(Int.min)` traps — surface it as a typed error instead.
            guard i != Int.min else {
                throw ExpressionError.evaluate("filter 'abs' cannot represent the absolute value of \(i)")
            }
            return .int(Swift.abs(i))
        case let .double(d): return .double(Swift.abs(d))
        default: throw ExpressionError.evaluate("filter 'abs' requires a number, got \(input.typeName)")
        }
    }

    /// `number | round(ndigits=0)` — returns a double (Jinja semantics).
    static let round: ExpressionFilter = { input, args in
        guard let value = input.asDouble else {
            throw ExpressionError.evaluate("filter 'round' requires a number, got \(input.typeName)")
        }
        let digits: Int
        switch args.count {
        case 0: digits = 0
        case 1:
            guard let d = args[0].intValue else {
                throw ExpressionError.evaluate("filter 'round' precision must be an integer")
            }
            digits = d
        default:
            throw ExpressionError.evaluate("filter 'round' expects 0 or 1 argument(s), got \(args.count)")
        }
        let factor = pow(10.0, Double(digits))
        return .double((value * factor).rounded() / factor)
    }

    // MARK: - Structured filters (transform DSL replacements)

    /// `object | merge(other)` — shallow merge; keys in `other` win.
    static let merge: ExpressionFilter = { input, args in
        try requireArgs(args, 1, "merge")
        var base = try requireObject(input, "merge")
        let other = try requireObject(args[0], "merge")
        for (k, v) in other { base[k] = v }
        return .object(base)
    }

    /// `object | pick('a', 'b')` — keep only the listed keys that are present.
    static let pick: ExpressionFilter = { input, args in
        let source = try requireObject(input, "pick")
        let keys = try requireKeyArgs(args, "pick")
        var result: [String: JinjaValue] = [:]
        for key in keys where source[key] != nil {
            result[key] = source[key]
        }
        return .object(result)
    }

    /// `object | omit('a', 'b')` — drop the listed keys.
    static let omit: ExpressionFilter = { input, args in
        var source = try requireObject(input, "omit")
        for key in try requireKeyArgs(args, "omit") {
            source.removeValue(forKey: key)
        }
        return .object(source)
    }

    /// `array | concat(other)` or `string | concat(other)` — concatenation.
    static let concat: ExpressionFilter = { input, args in
        try requireArgs(args, 1, "concat")
        switch (input, args[0]) {
        case let (.array(a), .array(b)):
            return .array(a + b)
        case let (.string(a), .string(b)):
            return .string(a + b)
        default:
            throw ExpressionError.evaluate(
                "filter 'concat' requires two arrays or two strings, got \(input.typeName) and \(args[0].typeName)"
            )
        }
    }
}

File: ts/src/filters.ts
// filters.ts — port of Expression/FilterRegistry.swift and Expression/Filters.swift
//
// The standard filter allowlist — the entire callable surface of the dialect.
// Each filter is type-checked and throws a typed ExpressionError on misuse.

import { ExpressionError } from "./errors.ts";
import {
	asDouble,
	type JinjaValue,
	JV,
	orderedAscending,
	orderedDescending,
	renderedString,
	semanticCompare,
	typeName,
} from "./value.ts";

export type ExpressionFilter = (input: JinjaValue, args: JinjaValue[]) => JinjaValue;

// MARK: - Argument helpers (function declarations — hoisted)

function requireArgs(args: JinjaValue[], count: number, name: string): void {
	if (args.length !== count) {
		throw ExpressionError.evaluate(
			`filter '${name}' expects ${count} argument(s), got ${args.length}`,
		);
	}
}

function requireString(value: JinjaValue, name: string): string {
	if (value.kind !== "string") {
		throw ExpressionError.evaluate(`filter '${name}' requires a string, got ${typeName(value)}`);
	}
	return value.value;
}

function requireObject(value: JinjaValue, name: string): Map<string, JinjaValue> {
	if (value.kind !== "object") {
		throw ExpressionError.evaluate(`filter '${name}' requires an object, got ${typeName(value)}`);
	}
	return value.value;
}

function requireKeyArgs(args: JinjaValue[], name: string): string[] {
	return args.map((a) => requireString(a, name));
}

function extreme(input: JinjaValue, name: string, wanted: number): JinjaValue {
	if (input.kind !== "array") {
		throw ExpressionError.evaluate(`filter '${name}' requires an array, got ${typeName(input)}`);
	}
	if (input.value.length === 0) {
		throw ExpressionError.evaluate(`filter '${name}' requires a non-empty array`);
	}
	let best = input.value[0]!;
	for (let i = 1; i < input.value.length; i++) {
		const item = input.value[i]!;
		if (semanticCompare(item, best) === wanted) best = item;
	}
	return best;
}

function roundHalfAwayFromZero(x: number): number {
	return x < 0 ? -Math.round(-x) : Math.round(x);
}

// MARK: - The standard filters

export const standardFilters: Record<string, ExpressionFilter> = {
	/** `value | default(fallback)` — fallback when the input is null. */
	default(input, args) {
		requireArgs(args, 1, "default");
		if (input.kind === "null") return args[0]!;
		return input;
	},

	/** `value | length` — element/character count of array, object, or string. */
	length(input, args) {
		requireArgs(args, 0, "length");
		switch (input.kind) {
			case "array":
				return JV.int(input.value.length);
			case "object":
				return JV.int(input.value.size);
			case "string":
				// Swift's String.count counts grapheme clusters; for the ASCII/BMP
				// inputs the dialect handles, code-point length agrees. Array.from
				// counts code points (closer to Swift than `.length`'s UTF-16 units).
				return JV.int(Array.from(input.value).length);
			default:
				throw ExpressionError.evaluate(
					`filter 'length' requires an array, object, or string, got ${typeName(input)}`,
				);
		}
	},

	/** `array | min`. */
	min(input, args) {
		requireArgs(args, 0, "min");
		return extreme(input, "min", orderedAscending);
	},

	/** `array | max`. */
	max(input, args) {
		requireArgs(args, 0, "max");
		return extreme(input, "max", orderedDescending);
	},

	lower(input, args) {
		requireArgs(args, 0, "lower");
		return JV.string(requireString(input, "lower").toLowerCase());
	},
	upper(input, args) {
		requireArgs(args, 0, "upper");
		return JV.string(requireString(input, "upper").toUpperCase());
	},
	trim(input, args) {
		requireArgs(args, 0, "trim");
		// Swift trims whitespacesAndNewlines; JS String.trim removes the same set
		// of ASCII/Unicode whitespace plus line terminators.
		return JV.string(requireString(input, "trim").trim());
	},

	/** `array | join(sep)` — render elements to strings and join (default ""). */
	join(input, args) {
		if (input.kind !== "array") {
			throw ExpressionError.evaluate(`filter 'join' requires an array, got ${typeName(input)}`);
		}
		let sep: string;
		if (args.length === 0) sep = "";
		else if (args.length === 1) sep = requireString(args[0]!, "join");
		else {
			throw ExpressionError.evaluate(
				`filter 'join' expects 0 or 1 argument(s), got ${args.length}`,
			);
		}
		return JV.string(input.value.map(renderedString).join(sep));
	},

	/** `number | abs`. */
	abs(input, args) {
		requireArgs(args, 0, "abs");
		if (input.kind === "int") return JV.int(Math.abs(input.value));
		if (input.kind === "double") return JV.double(Math.abs(input.value));
		throw ExpressionError.evaluate(`filter 'abs' requires a number, got ${typeName(input)}`);
	},

	/** `number | round(ndigits=0)` — returns a double (Jinja semantics). */
	round(input, args) {
		const value = asDouble(input);
		if (value === undefined) {
			throw ExpressionError.evaluate(`filter 'round' requires a number, got ${typeName(input)}`);
		}
		let digits: number;
		if (args.length === 0) digits = 0;
		else if (args.length === 1) {
			const a = args[0]!;
			if (a.kind !== "int") {
				throw ExpressionError.evaluate("filter 'round' precision must be an integer");
			}
			digits = a.value;
		} else {
			throw ExpressionError.evaluate(
				`filter 'round' expects 0 or 1 argument(s), got ${args.length}`,
			);
		}
		const factor = Math.pow(10, digits);
		// Swift's Double.rounded() is round-half-away-from-zero; match it.
		return JV.double(roundHalfAwayFromZero(value * factor) / factor);
	},

	/** `object | merge(other)` — shallow merge; keys in `other` win. */
	merge(input, args) {
		requireArgs(args, 1, "merge");
		const base = new Map(requireObject(input, "merge"));
		const other = requireObject(args[0]!, "merge");
		for (const [k, v] of other) base.set(k, v);
		return { kind: "object", value: base };
	},

	/** `object | pick('a','b')` — keep only listed keys that are present. */
	pick(input, args) {
		const source = requireObject(input, "pick");
		const keys = requireKeyArgs(args, "pick");
		const result = new Map<string, JinjaValue>();
		for (const key of keys) {
			if (source.has(key)) result.set(key, source.get(key)!);
		}
		return { kind: "object", value: result };
	},

	/** `object | omit('a','b')` — drop listed keys. */
	omit(input, args) {
		const source = new Map(requireObject(input, "omit"));
		for (const key of requireKeyArgs(args, "omit")) source.delete(key);
		return { kind: "object", value: source };
	},

	/** `array | concat(other)` or `string | concat(other)`. */
	concat(input, args) {
		requireArgs(args, 1, "concat");
		const other = args[0]!;
		if (input.kind === "array" && other.kind === "array") {
			return JV.array([...input.value, ...other.value]);
		}
		if (input.kind === "string" && other.kind === "string") {
			return JV.string(input.value + other.value);
		}
		throw ExpressionError.evaluate(
			`filter 'concat' requires two arrays or two strings, got ${typeName(input)} and ${typeName(other)}`,
		);
	},
};

// MARK: - Registry (declared after standardFilters to avoid TDZ on the static)

/** An immutable, allowlisted set of filters. */
export class FilterRegistry {
	private readonly filters: Map<string, ExpressionFilter>;

	constructor(filters: Record<string, ExpressionFilter> | Map<string, ExpressionFilter> = {}) {
		this.filters =
			filters instanceof Map ? new Map(filters) : new Map(Object.entries(filters));
	}

	filter(name: string): ExpressionFilter | undefined {
		return this.filters.get(name);
	}

	/** The set of filter names. */
	get names(): Set<string> {
		return new Set(this.filters.keys());
	}

	static readonly standard: FilterRegistry = new FilterRegistry({
		// Scalar / collection
		default: standardFilters.default!,
		length: standardFilters.length!,
		min: standardFilters.min!,
		max: standardFilters.max!,
		lower: standardFilters.lower!,
		upper: standardFilters.upper!,
		trim: standardFilters.trim!,
		join: standardFilters.join!,
		abs: standardFilters.abs!,
		round: standardFilters.round!,
		// Structured (transform DSL replacements)
		merge: standardFilters.merge!,
		pick: standardFilters.pick!,
		omit: standardFilters.omit!,
		concat: standardFilters.concat!,
	});
}
